firedrake/slate/slac/kernel_builder.py

```diff
@@ -42,8 +42,18 @@ def __init__(self, expression, tsfc_parameters=None):
         self.finalized_ast = None
         self._is_finalized = False
 
-        # Generate coefficient map (both mixed and non-mixed cases handled)
-        self.coefficient_map = prepare_coefficients(expression)
+        coefficient_map = OrderedDict()
+        for i, coefficient in enumerate(self.expression.coefficients()):
+            if type(coefficient.ufl_element()) == MixedElement:
+                csym_info = []
+                for j, _ in enumerate(coefficient.split()):
+                    csym_info.append(ast.Symbol("w_%d_%d" % (i, j)))
+            else:
+                csym_info = (ast.Symbol("w_%d" % i),)
+
+            coefficient_map[coefficient] = tuple(csym_info)
+
+        self.coefficient_map = coefficient_map
```

Comment:
If you don't want to pollute the constructor with this code, but don't want the `prepare_coefficients` free function (I agree it only appears here, so it makes sense to inline), you could make `coefficient_map` a `@cached_property` or `@property`, probably the former.